cumulus/client/cirrus-executor/src/processor.rs

```diff
@@ -0,0 +1,123 @@
+use sc_client_api::BlockBackend;
+use sp_api::ProvideRuntimeApi;
+use sp_runtime::{generic::BlockId, traits::Block as BlockT};
+
+use cirrus_node_primitives::ProcessorResult;
+use cirrus_primitives::{AccountId, SecondaryApi};
+use sp_executor::{ExecutionReceipt, OpaqueBundle};
+use subspace_runtime_primitives::Hash as PHash;
+
+use super::{Executor, LOG_TARGET};
+use codec::{Decode, Encode};
+
+impl<Block, BS, RA, Client, TransactionPool> Executor<Block, BS, RA, Client, TransactionPool>
+where
+	Block: BlockT,
+	Client: sp_blockchain::HeaderBackend<Block>,
+	BS: BlockBackend<Block>,
+	RA: ProvideRuntimeApi<Block>,
+	RA::Api: SecondaryApi<Block, AccountId>,
+	TransactionPool: sc_transaction_pool_api::TransactionPool<Block = Block>,
+{
+	/// Actually implements `process_bundles`.
+	pub(super) async fn process_bundles_impl(
+		self,
+		primary_hash: PHash,
+		bundles: Vec<OpaqueBundle>,
+	) -> Option<ProcessorResult> {
+		// TODO:
+		// 1. [x] convert the bundles to a full tx list
+		// 2. [x] duplicate the full tx list
+		// 3. shuffle the full tx list by sender account
+		let mut extrinsics = bundles
+			.into_iter()
+			.map(|bundle| {
+				bundle.opaque_extrinsics.into_iter().filter_map(|opaque_extrinsic| {
+					match <<Block as BlockT>::Extrinsic>::decode(
+						&mut opaque_extrinsic.encode().as_slice(),
+					) {
+						Ok(uxt) => Some(uxt),
+						Err(e) => {
+							tracing::error!(
+								target: LOG_TARGET,
+								error = ?e,
+								"Failed to decode the opaque extrisic in bundle, this should not happen"
+							);
+							None
+						},
+					}
+				})
+			})
+			.flatten()
+			.collect::<Vec<_>>();
+
+		// TODO: or just Vec::new()?
+		// Ideally there should be only a few duplicated transactions.
+		let mut seen = Vec::with_capacity(extrinsics.len());
```

Comment:
`HashSet` is a better fit for this purpose

Reply:
`Block::Extrinsic` doesn't have the `Hash` bound so `HashSet` can't be used.